common/serde-reflection/src/trace.rs

```diff
@@ -2,473 +2,233 @@
-}
-
-impl<'a> ser::SerializeTupleVariant for TupleVariantTracer<'a> {
-    type Ok = Format;
-    type Error = Error;
 
-    fn serialize_field<T>(&mut self, value: &T) -> Result<()>
+    /// Compute a default value for `T`.
+    /// Return true if `T` is an enum and more values are needed to cover
+    /// all variant cases. Otherwise similar to `trace_type_once`.
+    pub fn sample_type_once<'de, T>(&mut self) -> Result<(T, bool)>
     where
-        T: ?Sized + Serialize,
+        T: Deserialize<'de>,
     {
-        let format = value.serialize(&mut *self.tracer)?;
-        self.fields.push(format);
-        Ok(())
-    }
-
-    fn end(self) -> Result<Format> {
-        let variant = VariantFormat::Tuple(self.fields);
-        self.tracer
-            .record_variant(self.name, self.variant_index, self.variant_name, variant)
-    }
-}
-
-pub struct MapTracer<'a> {
-    tracer: &'a mut Tracer,
-    key: Format,
-    value: Format,
-}
-
-impl<'a> ser::SerializeMap for MapTracer<'a> {
-    type Ok = Format;
-    type Error = Error;
-
-    fn serialize_key<T>(&mut self, key: &T) -> Result<()>
+        let mut format = Format::Unknown;
+        let deserializer = Deserializer::new(&mut *self, &mut format);
+        let value = T::deserialize(deserializer)?;
+        let has_more = match &format {
+            Format::TypeName(name) => {
+                if self.incomplete_enums.contains(name) {
+                    self.incomplete_enums.remove(name);
+                    true
+                } else {
+                    false
+                }
+            }
+            _ => false,
+        };
+        Ok((value, has_more))
+    }
+
+    /// Same as sample_type_once but if `T` is an enum, we repeat the process
+    /// until all variants are covered. Otherwise similar to `trace_type`.
+    pub fn sample_type<'de, T>(&mut self) -> Result<Vec<T>>
     where
-        T: ?Sized + Serialize,
+        T: Deserialize<'de>,
     {
-        let format = key.serialize(&mut *self.tracer)?;
-        self.key.merge(format)
+        let mut values = Vec::new();
+        loop {
+            let (value, has_more) = self.sample_type_once::<T>()?;
+            values.push(value);
+            if !has_more {
+                return Ok(values);
+            }
+        }
     }
 
-    fn serialize_value<T>(&mut self, value: &T) -> Result<()>
-    where
-        T: ?Sized + Serialize,
-    {
-        let format = value.serialize(&mut *self.tracer)?;
-        self.value.merge(format)
+    /// Finish tracing and recover a map of normalized formats.
+    /// Returns an error if we detect incompletely traced types.
+    /// This may happen in a few of cases:
+    /// * We traced serialization only and the sampled values missed some variant, the content
+    ///   of an option type, the content of a sequence type, the key or the value of a dictionary type.
+    /// * We trace deserialization of some types but not every enum type was explicitly traced.
```

Comment:
These are both cases where a library user could remedy the problem by providing more types right?

Reply:
Yes, that's correct. The main doc has more info now.